### PR TITLE
Expose serialized data for transaction signatures

### PR DESCRIPTION
**Pull request by ipaljak-tbtl**

### CODE REVIEW

File: src/util/bip143.rs [hunk 1]
```diff
@@ -168,15 +169,20 @@ impl<R: Deref<Target=Transaction>> SigHashCache<R> {
         })
     }
 
-    /// Compute the BIP143 sighash for any flag type. See SighashComponents::sighash_all simpler
-    /// API for the most common case
-    pub fn signature_hash(&mut self, input_index: usize, script_code: &Script, value: u64, sighash_type: SigHashType) -> SigHash {
-
+    /// Encode the BIP143 signing data for any flag type into a given object implementing a
+    /// std::io::Write trait.
+    pub fn encode_signing_data<Write: io::Write>(
```
Comment (dpc) [inline src/util/bip143.rs:174]: I'm not sure if `Write` should be first or last (which one is more idiomatic).

Also `enc` is not a best name`. Just `writer` would be better, IMO.

Maybe a `encode_signing_data_to` would be a better name. Maybe. Not sure.

Comment (ipaljak-tbtl) [reply src/util/bip143.rs:174]: I agree on both of your suggestions regarding naming (`enc` -> `writer` and `encode_signing_data` -> `encode_signing_data_to`).

I'm not sure what is more idiomatic regarding the order of function parameters, but it makes more sense to me that `writer` becomes the first parameter if we rename the function. 

All in all, I'm fine with accepting all of your suggestions.
Thanks!

Comment (dpc) [reply src/util/bip143.rs]: Looks OK to me now. I'm more of a lurker around here, so you're going to need some more authoritative review.

**Review by dr-orlovsky — APPROVED** [review]
LGTM

**Comment by apoelstra** [comment]
Can you squash these commits?

File: src/blockdata/transaction.rs [hunk 2]
```diff
@@ -307,30 +307,34 @@ impl Transaction {
...
+    /// # Panics Panics if `input_index` is greater than or equal to `self.input.len()`
     ///
-    pub fn signature_hash(&self, input_index: usize, script_pubkey: &Script, sighash_u32: u32) -> SigHash {
+    pub fn encode_signing_data_to<Write: io::Write>(
+        &self,
+        mut writer: Write,
+        input_index: usize,
+        script_pubkey: &Script,
+        sighash_u32: u32
+    ) {
```
Comment (apoelstra) [inline src/blockdata/transaction.rs:327]: This should return a `Result<(), io::Error>` (or a `io::Result<()>` if that makes you happy)

Comment (apoelstra) [reply src/blockdata/transaction.rs:327]: Actually it needs to return a `Result<(), encode::Error>`

File: src/blockdata/transaction.rs [hunk 3]
```diff
@@ -307,30 +307,34 @@ impl Transaction {
...
         // Special-case sighash_single bug because this is easy enough.
         if sighash == SigHashType::Single && input_index >= self.output.len() {
-            return SigHash::from_slice(&[1, 0, 0, 0, 0, 0, 0, 0,
-                                               0, 0, 0, 0, 0, 0, 0, 0,
-                                               0, 0, 0, 0, 0, 0, 0, 0,
-                                               0, 0, 0, 0, 0, 0, 0, 0]).unwrap();
+            writer.write_all(&[1, 0, 0, 0, 0, 0, 0, 0,
+                               0, 0, 0, 0, 0, 0, 0, 0,
+                               0, 0, 0, 0, 0, 0, 0, 0,
+                               0, 0, 0, 0, 0, 0, 0, 0]).unwrap();
```
Comment (apoelstra) [inline src/blockdata/transaction.rs:337]: This `unwrap()`, and all `unwrap()`s, should be `?`s.

Also here the old code did an early return; you need to add `return Ok(())` to preserve this behavior.

File: src/util/bip143.rs [hunk 4]
```diff
@@ -168,53 +169,65 @@ impl<R: Deref<Target=Transaction>> SigHashCache<R> {
...
+    /// Encode the BIP143 signing data for any flag type into a given object implementing a
+    /// std::io::Write trait.
+    pub fn encode_signing_data_to<Write: io::Write>(
+        &mut self,
+        mut writer: Write,
+        input_index: usize,
+        script_code: &Script,
+        value: u64,
+        sighash_type: SigHashType,
+    ) {
```
Comment (apoelstra) [inline src/util/bip143.rs:181]: Ditto here; return an error, use `?` rather than unwrapping

**Review by apoelstra — APPROVED** [review]
utACK e66caab956ff003bba3f15b0fcb08944d5aacd3c

This is great! Happy to see the number of `unwrap`s in this lib go down by so much.

**Review by sgeisler — APPROVED** [review]
LGTM

**Comment by dr-orlovsky** [comment]
This one requires `API break` label to be applied b/c of https://github.com/rust-bitcoin/rust-miniscript/pull/153

**Comment by sgeisler** [comment]
@dr-orlovsky are you sure it's not just an incompatibility with rust `1.22` (which we don't guarantee anymore since `0.25`)?

File: src/util/bip143.rs [hunk 5]
```diff
@@ -168,54 +169,72 @@ impl<R: Deref<Target=Transaction>> SigHashCache<R> {
...
+
+    /// Compute the BIP143 sighash for any flag type. See SighashComponents::sighash_all simpler
+    /// API for the most common case
+    pub fn signature_hash(
+        &mut self,
+        input_index: usize,
+        script_code: &Script,
+        value: u64,
+        sighash_type: SigHashType
+    ) -> Result<SigHash, encode::Error> {
```
Comment (apoelstra) [inline src/util/bip143.rs:234]: This API should not have changed.

Comment (apoelstra) [reply src/util/bip143.rs:234]: #493 fixes this, all good now. Thanks!!

**Pull request merged**